src/rust/cryptography-x509/src/certificate.rs

```diff
@@ -14,6 +27,52 @@ pub struct Certificate<'a> {
     pub signature: asn1::BitString<'a>,
 }
 
+impl Certificate<'_> {
+    /// Returns the certificate's issuer.
+    pub fn issuer(&self) -> &NameReadable<'_> {
+        self.tbs_cert.issuer.unwrap_read()
+    }
+
+    /// Returns the certificate's subject.
+    pub fn subject(&self) -> &NameReadable<'_> {
+        self.tbs_cert.subject.unwrap_read()
+    }
+
+    /// Returns whether the certificate is "self-issued", whether its
+    /// issuer and subject are the same.
+    pub fn is_self_issued(&self) -> bool {
+        self.issuer() == self.subject()
+    }
+
+    /// Returns an iterable container over the certificate's extension, or
+    /// an error if the extension set contains a duplicate extension.
+    pub fn extensions(&self) -> Result<Extensions<'_>, CertificateError> {
+        self.tbs_cert
+            .extensions()
+            .map_err(CertificateError::DuplicateExtension)
+    }
+
+    /// Returns whether the given extension (by OID) is critical, or
+    /// false if the extension is not present.
+    pub fn extension_is_critical(&self, oid: &asn1::ObjectIdentifier) -> bool {
```

Comment:
This feels pretty ugly to me, but it avoids a bit of combinator juggling in the validator APIs, e.g.:

```rust
        if cert.extension_is_critical(&SUBJECT_DIRECTORY_ATTRIBUTES_OID) {
            return Err("SubjectDirectoryAttributes must not be marked critical".into());
        }
```

...rather than having to poke through `extension()?` and then handling the `None` case.

Reply:
To me this seems too specific, and will lead to way too much iteration over extensions. The right way to do this API is to get the extension, check its criticality,then parse it.